### PR TITLE
✨ Added migration of cloud_for_customers.py source to 2.0

### PR DESCRIPTION
**Pull request by fgoiriz**
✨ Added migration for cloud_for_customers.py source

Changed docstring and credentials handling

### CODE REVIEW

File: viadot/sources/cloud_for_customers.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        **kwargs,
+    ):
+
+        ## Credentials logic
+        credentials = credentials or get_source_credentials(config_key)
+        if credentials is None:
+            raise CredentialError("Please specify the credentials.")
+        CloudForCustomersCredentials(**credentials)  # validate the credentials schema
+        super().__init__(*args, credentials=credentials, **kwargs)
+        self.logger.info(credentials)
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:56]: Pls remove

**Review by Trymzet — CHANGES_REQUESTED** [review]
Overall looks good but you didn't fix the docstrings to match the formatting we use, and also didn't improve any language (descriptions etc.). See `sharepoint.py` for examples:
```python
        """
        Download a file from Sharepoint.
        Args:
            url (str): The URL of the file to be downloaded.
            to_path (str): Where to download the file.
        Example:
            download_file(
                url="https://{tenant_name}.sharepoint.com/sites/{directory}/Shared%20Documents/Dashboard/file",
                to_path="file.xlsx"
            )
        """
```

vs one of yours:
```python
        """Fetches the data from source with url.
        At first enter url is a join of url and endpoint passed into this function.
        At any other entering it bring `__next_url` adress, generated automatically, but without params.
        """
```
Can you spot the differences? :) 

Also, the language is unclear in a lot of places so this should also be improved. For example, I don't know what "At first enter url is a join of url and endpoint passed into this function." even means.

**Review by Trymzet — COMMENTED** [review]
This looks good, now we just need tests, at a minimum we need a test for the `to_df()` method.

File: viadot/sources/cloud_for_customers.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        pw = self.credentials.get("password")
+        response = handle_api_response(
+            url=url,
+            params=params,
+            auth=(username, pw),
+            timeout=timeout,
+        )
+        return response
+
+    def records_to_df(
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:230]: Rename to `to_df()`

**Review by Trymzet — CHANGES_REQUESTED** [review]

File: viadot/sources/cloud_for_customers.py (hunk 2 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py]: Done

File: viadot/utils.py [hunk 3]
```diff
@@ -18,6 +18,42 @@
 from .exceptions import APIError
 from .signals import SKIP
 
+import json
+
+from prefect_azure import AzureKeyVaultSecretReference
+
+
+def get_credentials(credentials_block_name: str) -> dict:
```
Comment (Trymzet) [inline viadot/utils.py:26]: This should not be here as viadot is decoupled from `prefect`.

File: tests/test_cloud_for_customers.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+cursor_field = "EntityLastChangedOn"
+cursor_value = "2022-10-30T00:00:00Z"
+cursor_filter = f"{cursor_field} ge datetimeoffset'{cursor_value}'"
+filter_params = {"$filter": cursor_filter}
+
+
+def test_cloud_for_customers_to_df():
+    credentials_secret = "aia-c4c-qa"
+    credentials = get_credentials(credentials_secret)
```
Comment (Trymzet) [inline tests/test_cloud_for_customers.py:19]: You need to use viadot config here and the `get_source_credentials()` util

File: tests/test_cloud_for_customers.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+"""Tests for CloudForCustomers source class"""
+
+from viadot.utils import get_credentials
+from viadot.sources.cloud_for_customers import CloudForCustomers
+import pandas as pd
+
+
+url = "https://my336539.crm.ondemand.com/sap/c4c/odata/v1/c4codataapi/"
```
Comment (Trymzet) [inline tests/test_cloud_for_customers.py:8]: This should not be provided here but put in `conftest` similar as other sensitive settings.

File: viadot/sources/cloud_for_customers.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+from viadot.exceptions import CredentialError
+
+from ..utils import handle_api_response
+from ..config import get_source_credentials
+from .base import Source
+
+
+class CloudForCustomersCredentials(BaseModel):
+    username: str  # CloudForCustomers username (e.g username@{tenant_name}.com).
+    password: str  # CloudForCustomers password.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:18]: I don't this these comments are necessary, maybe besides the example which could just read "# eg. username@{tenant_name}.com"

File: viadot/sources/cloud_for_customers.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+
+class CloudForCustomers(Source):
+    """Cloud for Customers connector to fetch Odata source.
+
+    Args:
+        url (str, optional): The URL to the C4C API. E.g 'https://my336539.crm.ondemand.com/sap/c4c/odata/v1/c4codataapi/'.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:25]: Don't provide our private instance as an example

File: viadot/sources/cloud_for_customers.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        endpoint (str, optional): The API endpoint.
+        report_url (str, optional): The API url in case of prepared report.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:27]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+        filter_params (Dict[str, Any], optional): Filtering parameters passed to the request. E.g {"$filter": "AccountID eq '1234'"}. More info: https://userapps.support.sap.com/sap/support/knowledge/en/2330688
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:28]: this line is too long, can you pls split it into two?

File: viadot/sources/cloud_for_customers.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+        ## Credentials logic
+        credentials = credentials or get_source_credentials(config_key)
+        if credentials is None:
+            raise CredentialError("Please specify the credentials.")
+        CloudForCustomersCredentials(**credentials)  # validate the credentials schema
+        super().__init__(*args, credentials=credentials, **kwargs)
+        ## End Credentials logic
+
+        self.url = url or self.credentials.get("server")
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:55]: Can you change the config name too since it's not really a server? Isn't it a URL or more specifically, the API endpoint?

File: viadot/sources/cloud_for_customers.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+            self.filter_params = filter_params_merged
+        else:
+            self.filter_params = self.DEFAULT_PARAMS
+
+        if self.url:
+            self.full_url = urljoin(self.url, self.query_endpoint)
+
+    @staticmethod
+    def create_metadata_url(url: str) -> str:
+        """Creates url to fetch metadata to.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:77]: Did you mean from? Also I don't understand this explanation, it feels like half of the information.

File: viadot/sources/cloud_for_customers.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+        Args:
+            url (str): The url to transform to metadata url.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:80]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+        Returns:
+            meta_url (str): The url to fetch metadata with.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:83]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+        """
+        start = url.split(".svc")[0]
+        url_raw = url.split("?")[0]
+        end = url_raw.split("/")[-1]
+        meta_url = start + ".svc/$metadata?entityset=" + end
+        return meta_url
+
+    def _extract_records_from_report_url(self, report_url: str) -> List[Dict[str, Any]]:
+        """Fetches report_url to exctract records.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:92]: typo

File: viadot/sources/cloud_for_customers.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+
+        Args:
+            report_url (str): The url to extract records from.
+
+        Returns:
+            records (List[Dict[str, Any]]): The records extracted from url.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:98]: url -> URL (or say "... from `report_url`")

File: viadot/sources/cloud_for_customers.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+    def _extract_records_from_url(self, url: str) -> List[Dict[str, Any]]:
+        """Fetches url to exctract records.
+
+        Args:
+            url (str): The url to extract records from.
+
+        Returns:
+            records (List[Dict[str, Any]]): The records extracted from url.
+        """
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:119]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+            if isinstance(response_json["d"], dict):
+                # ODATA v2+ API
+                new_records = response_json["d"].get("results")
+                url = response_json["d"].get("__next", None)
+            else:
+                # ODATA v1
+                new_records = response_json["d"]
+                url = response_json.get("__next", None)
+
+            # prevents concatenation of previous url's with filter_params with the same filter_params
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:135]: url's -> urls

File: viadot/sources/cloud_for_customers.py [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+            tmp_filter_params = None
+            tmp_full_url = url
+
+            if hasattr(new_records, "__iter__"):
+                records.extend(new_records)
+        return records
+
+    def extract_records(self) -> List[Dict[str, Any]]:
+        """Downloads records from url or report_url if present.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:144]: If we're referencing variables I think it's more clear to use the backticks

File: viadot/sources/cloud_for_customers.py [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+
+        Returns:
+            records (List[Dict[str, Any]]): The records extracted from url.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:147]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+            return self._extract_records_from_url(url=self.full_url)
+
+    def get_entities(
+        self, dirty_json: Dict[str, Any], url: str
+    ) -> List[Dict[str, Any]]:
+        """Extracts entities from request.json().
+
+        Args:
+            dirty_json (Dict[str, Any]): request.json() dict from response to API.
+            url (str): The url to fetch metadata from.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:161]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+                        else:
+                            new_entity[key] = object_of_interest
+            entities.append(new_entity)
+        return entities
+
+    def get_mapping_property_to_sap_level(self, url: str = None) -> Dict[str, str]:
+        """Creates Dict mapping property Name to value of sap label.
+
+        Args:
+            url (str, optional): The url to fetch metadata from.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:187]: url -> URL

File: viadot/sources/cloud_for_customers.py [hunk 22]
```diff
@@ -0,0 +1,256 @@
...
+
+        Returns:
+            Dict[str, str]: Property Name to value of sap label Dict.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:190]: sap -> SAP; what's the "Dict" at the end?

File: viadot/utils.py [hunk 23]
```diff
@@ -103,8 +141,8 @@ def handle_api_response(
     response = get_response(
         url=url, auth=auth, params=params, headers=headers, timeout=timeout
     )
-    response_handled = handle_response(response)
 
+    response_handled = handle_response(response, url)
```
Comment (Trymzet) [inline viadot/utils.py:145]: Can't you take the url from the Response object? At any rate I don't think you should require another param just to log some details. Use the Response object to get any relevant info or re-raise exceptions.

**Review by Trymzet — CHANGES_REQUESTED** [review]
Couple comments

File: viadot/sources/cloud_for_customers.py (hunk 22 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:190]: The sentence is basically: "x to y dictionary"
x being "Property Name"
y being "value of sap label"

Comment (Trymzet) [reply viadot/sources/cloud_for_customers.py:190]: what is SAP label dict then?

File: viadot/utils.py (hunk 3 above)
Comment (fgoiriz) [reply viadot/utils.py:26]: Removed

File: tests/test_cloud_for_customers.py (hunk 4 above)
Comment (fgoiriz) [reply tests/test_cloud_for_customers.py:19]: Done

File: tests/test_cloud_for_customers.py (hunk 5 above)
Comment (fgoiriz) [reply tests/test_cloud_for_customers.py:8]: Done

File: viadot/sources/cloud_for_customers.py (hunk 6 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:18]: Done

File: viadot/sources/cloud_for_customers.py (hunk 7 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:25]: Fixed

File: viadot/sources/cloud_for_customers.py (hunk 9 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:28]: Done

File: viadot/sources/cloud_for_customers.py (hunk 11 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:77]: Yes, typo, I meant "from".
Fixed

File: viadot/sources/cloud_for_customers.py (hunk 13 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:83]: Done

File: viadot/sources/cloud_for_customers.py (hunk 14 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:92]: Fixed

File: viadot/sources/cloud_for_customers.py (hunk 15 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:98]: replaced url with report_url

File: viadot/sources/cloud_for_customers.py (hunk 16 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:119]: Done

File: viadot/sources/cloud_for_customers.py (hunk 17 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:135]: Done

File: viadot/sources/cloud_for_customers.py (hunk 18 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:144]: Done

File: viadot/sources/cloud_for_customers.py (hunk 19 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:147]: Done

File: viadot/sources/cloud_for_customers.py (hunk 20 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:161]: Done

File: viadot/sources/cloud_for_customers.py (hunk 21 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:187]: Done

File: viadot/sources/cloud_for_customers.py (hunk 22 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:190]: "SAP lebel" is not a dict, it is a str, "property value" also is a str.
the "Dict" is not referring to "SAP label" a dict, it is referring to the object to be returned, which is a "x to y Dict"
but maybe this is confusing, I will remove the "Dict"

File: viadot/utils.py (hunk 23 above)
Comment (fgoiriz) [reply viadot/utils.py:145]: Done

File: viadot/sources/cloud_for_customers.py (hunk 10 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:55]: Ok, I change it in the source, to search for "url", but then we would have to change the config block to have "url" as a key

File: tests/conftest.py [hunk 24]
```diff
@@ -8,6 +8,11 @@
 load_dotenv()
 
 
+@pytest.fixture(scope="session")
+def TEST_C4C_API_URL():
+    return "https://my336539.crm.ondemand.com/sap/c4c/odata/v1/c4codataapi/"
```
Comment (Trymzet) [inline tests/conftest.py:13]: Should be read from env as the endpoint is private

File: tests/test_cloud_for_customers.py [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+CONTACT_ENDPOINT = "ContactCollection"
+cursor_field = "EntityLastChangedOn"
+cursor_value = "2022-10-30T00:00:00Z"
+cursor_filter = f"{cursor_field} ge datetimeoffset'{cursor_value}'"
+filter_params = {"$filter": cursor_filter}
+
+
+def test_cloud_for_customers_to_df(TEST_C4C_API_URL):
+    credentials_secret = "aia-c4c-qa"
+    print(credentials_secret)
```
Comment (Trymzet) [inline tests/test_cloud_for_customers.py:17]: remove

File: tests/test_cloud_for_customers.py [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+    credentials = get_credentials(credentials_secret)
+    print(credentials)
```
Comment (Trymzet) [inline tests/test_cloud_for_customers.py:19]: remove... if we were running the tests in CI these credentials would already be compromised as they would show up in logs...

File: tests/test_cloud_for_customers.py [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+    c4c = CloudForCustomers(
+        url=TEST_C4C_API_URL,
+        endpoint=CONTACT_ENDPOINT,
+        filter_params=filter_params,
+        credentials=credentials,
+    )
+
+    df = c4c.to_df(fields=None, dtype=None)
+
+    assert isinstance(df, pd.core.frame.DataFrame)
```
Comment (Trymzet) [inline tests/test_cloud_for_customers.py:29]: isn't it just `pd.DataFrame`?

File: viadot/sources/cloud_for_customers.py [hunk 28]
```diff
@@ -0,0 +1,257 @@
...
+from viadot.exceptions import CredentialError
+
+from ..utils import handle_api_response
+from ..config import get_source_credentials
+from .base import Source
+
+
+class CloudForCustomersCredentials(BaseModel):
+    username: str  # eg. username@{tenant_name}.com
+    password: str
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:18]: maybe we should use SecretStr here? can you give me your opinion

File: viadot/sources/cloud_for_customers.py [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+
+
+class CloudForCustomers(Source):
+    """Cloud for Customers connector to fetch Odata source.
+
+    Args:
+        url (str, optional): The URL to the C4C API. E.g 'https://my396876543.jrt.ondemand.com/c4c/v1/'.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:25]: remove private instance

File: viadot/sources/cloud_for_customers.py [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+        endpoint (str, optional): The API endpoint.
+        report_url (str, optional): The API URL in case of prepared report.
+        filter_params (Dict[str, Any], optional): Filtering parameters passed to the request. E.g {"$filter": "AccountID eq '1234'"}.
+        More info on: https://userapps.support.sap.com/sap/support/knowledge/en/2330688
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:29]: tabulate this line

File: viadot/sources/cloud_for_customers.py [hunk 31]
```diff
@@ -0,0 +1,257 @@
...
+            self.filter_params = filter_params_merged
+        else:
+            self.filter_params = self.DEFAULT_PARAMS
+
+        if self.url:
+            self.full_url = urljoin(self.url, self.query_endpoint)
+
+    @staticmethod
+    def create_metadata_url(url: str) -> str:
+        """Creates url to fetch metadata from.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:78]: 1) Let's be consistent, and I think "URL" is the correct grammar (like you use it on line 81 for example).
2) I didn't get the explanation TBH, I still don't know what this method does. I only get what `meta_url` is.

File: viadot/sources/cloud_for_customers.py [hunk 32]
```diff
@@ -0,0 +1,257 @@
...
+        """
+        if self.is_report:
+            return self._extract_records_from_report_url(url=self.report_url)
+        else:
+            return self._extract_records_from_url(url=self.full_url)
+
+    def get_entities(
+        self, dirty_json: Dict[str, Any], url: str
+    ) -> List[Dict[str, Any]]:
+        """Extracts entities from request.json().
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:158]: What do you mean by entities, some C4C objects? Maybe a link to docs or an example entity could be useful here.

File: viadot/sources/cloud_for_customers.py [hunk 33]
```diff
@@ -0,0 +1,257 @@
...
+                        new_key = column_maper_dict.get(key)
+                        if new_key:
+                            new_entity[new_key] = object_of_interest
+                        else:
+                            new_entity[key] = object_of_interest
+            entities.append(new_entity)
+        return entities
+
+    def get_mapping_property_to_sap_level(self, url: str = None) -> Dict[str, str]:
+        """Creates Dict mapping property Name to value of sap label.
```
Comment (Trymzet) [inline viadot/sources/cloud_for_customers.py:185]: what for, can you add a line maybe for the why, as it seems like a random thing to do?

File: tests/test_cloud_for_customers.py (hunk 25 above)
Comment (fgoiriz) [reply tests/test_cloud_for_customers.py:17]: Done

File: tests/test_cloud_for_customers.py (hunk 26 above)
Comment (fgoiriz) [reply tests/test_cloud_for_customers.py:19]: True, sorry for this one. I will review my own PR's after pushing new changes to avoid completely this kind of silly mistaked.

File: viadot/sources/cloud_for_customers.py (hunk 28 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:18]: Great! will prevent some security issues like logging sensitive data.

File: viadot/sources/cloud_for_customers.py (hunk 29 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:25]: It is not private instance, it is a random example I typed that looks like an actual instance url.
Letters and numbers were chosen randomly, while preserving similar url structure to a real one.

File: viadot/sources/cloud_for_customers.py (hunk 30 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:29]: This is what I mean when I say assertive communication. It is not obvoius where do you want the tabulation and on which line of code. it is intuitable but not unambiguous. I would have to push a wrong tabulation or ask you and then wait for your answer to know exactly what change you want. Maybe im dumb and I can not recognize the obvious tabulation that is needed here (it is a possibility) but it does not matter, as a rule of thumb, communication should be foolproof

This could be avoided if the comment would have been something like: "tabulate line 29, between x and y" or something like that.

Hope you take this as constructive feedback because it was given with that intention.🙂

Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:29]: I will make the tabulation between : and https

File: viadot/sources/cloud_for_customers.py (hunk 31 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:78]: 1. Done
2. let us remember I didnt write this code so I am not fully 100% sure what this does.
But I think it is kind of clear that it is transforming the original url string, adding and removing stuff to the string, to change the composition of the url.
A simpler example containing the escence of this method would be:
def create_metadata_url(url: str = "https://mydomain.com/myendpoint"):
    meta_url = url + "metadata_endpoint/"
    return meta_url

Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:78]: I could be wrong on this one, but I think it is practically what I have explained

File: viadot/sources/cloud_for_customers.py (hunk 32 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:158]: I am not sure on this one, since this was already like this, and I found no documentation regarding this one.

Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:158]: I will try to guess this following days by printing these objects and trying to understand what entities mean. Also I will read some C4C Documentation, If you have some link to a useful C4C documentation it would be helpful.

File: viadot/sources/cloud_for_customers.py (hunk 33 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:185]: Same as the previous I am not sure, because this was already like this, and I found no documentation regarding the why of this method. I dont even know what mapping property or sap level is. But I will try to investigate about this things.

File: tests/test_cloud_for_customers.py [hunk 34]
```diff
@@ -0,0 +1,27 @@
+"""Tests for CloudForCustomers source class"""
+from viadot.config import get_source_credentials as get_credentials
```
Comment (Trymzet) [inline tests/test_cloud_for_customers.py:2]: I would not do that as there is a function with this name in `prefect-viadot`, that might be confusing?

**Review by Trymzet — COMMENTED** [review]
Added my comments

File: tests/conftest.py (hunk 24 above)
Comment (fgoiriz) [reply tests/conftest.py]: Done

File: tests/test_cloud_for_customers.py (hunk 34 above)
Comment (fgoiriz) [reply tests/test_cloud_for_customers.py:2]: Fixed

File: viadot/sources/cloud_for_customers.py (hunk 32 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:158]: Done

File: viadot/sources/cloud_for_customers.py (hunk 33 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py]: Added info in docstring.

File: tests/test_cloud_for_customers.py (hunk 27 above)
Comment (Trymzet) [reply tests/test_cloud_for_customers.py:26]: Can you change to `pd.DataFrame` then? It's typed like this everywhere else in the library so it would also be inconsistent with existing code aside from the verbosity.

File: viadot/sources/cloud_for_customers.py (hunk 29 above)
Comment (Trymzet) [reply viadot/sources/cloud_for_customers.py:25]: .

**Review by Trymzet — CHANGES_REQUESTED** [review]

File: tests/test_cloud_for_customers.py (hunk 27 above)
Comment (fgoiriz) [reply tests/test_cloud_for_customers.py:26]: Ok

File: viadot/sources/cloud_for_customers.py (hunk 29 above)
Comment (fgoiriz) [reply viadot/sources/cloud_for_customers.py:25]: Fixed

**Review by Trymzet — APPROVED** [review]
Looks good
